File: leetcode/22.generate-parentheses.py
#
# @lc app=leetcode id=22 lang=python3
#
# [22] Generate Parentheses
#

# @lc code=start
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:

        stack = []
        res = []

        def dfs(open, close):
            if open == close == n:
                res.append("".join(stack))

            if open < n:
                stack.append("(")
                dfs(open + 1, close)
                stack.pop()
            if close < open:
                stack.append(")")
                dfs(open, close + 1)
                stack.pop()

        dfs(0, 0)
        return res


# @lc code=end

File: leetcode/22.generate-parentheses.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/*
 * @lc app=leetcode id=22 lang=java
 *
 * [22] Generate Parentheses
 */

// @lc code=start
class Solution {
    List<String> res = new ArrayList<>();
    List<String> temp = new ArrayList<>();
    int n;

    public List<String> generateParenthesis(int n) {
        this.n = n;
        dfs(0, 0);
        return res;
    }

    public void dfs(int open, int close) {
        if (open == close && close == n) {
            res.add(String.join("", temp));
        }
        if (open < n) {
            temp.add("(");
            dfs(open + 1, close);
            temp.remove(temp.size() - 1);
        }
        if (close < open) {
            temp.add(")");
            dfs(open, close + 1);
            temp.remove(temp.size() - 1);
        }

    }
}
// @lc code=end
